### PR TITLE
Order

### PR DESCRIPTION
**Pull request by Mauricio-Natacci**

### CODE REVIEW

File: backend/database/models/order.model.ts [hunk 1]
```diff
@@ -0,0 +1,46 @@
+import { getModelForClass, modelOptions, prop, Ref } from '@typegoose/typegoose'
+import { User } from './user.model'
+
+@modelOptions({ options: { allowMixed: 0 } })
```
Comment (nathan-torquato) [inline backend/database/models/order.model.ts:4]: What's this `allowMixed` used for?

File: backend/graphql/resolvers/order.resolver.ts [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+  CreateOrderInput,
+  GetOrderInput,
+  OrderCanceledInput,
+  OrderDeliveredInput
+} from '../inputs/order.input'
+import { OrderService } from '../service/order.service'
+import { Context } from '../types/context.type'
+import { Order } from '../types/order.type'
+import { CreateOrderUseCase } from '../usecases/create-order.use-case'
+
```
Comment (nathan-torquato) [inline backend/graphql/resolvers/order.resolver.ts:14]: This file looks amazing 💎

File: backend/graphql/service/order.service.ts [hunk 3]
```diff
@@ -0,0 +1,62 @@
+/* eslint-disable @typescript-eslint/no-non-null-assertion */
+import { OrderModel } from '../../database/models/order.model'
+import { NotFoundError } from '../errors/notFoundError'
+import {
+  type OrderDeliveredInput,
+  type GetOrderInput,
+  type OrderCanceledInput
+} from '../inputs/order.input'
+import { type Context } from '../types/context.type'
```
Comment (nathan-torquato) [inline backend/graphql/service/order.service.ts:9]: why do you need the `{ type ...` part?

File: backend/graphql/service/order.service.ts [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+
+    return order
+  }
+
+  async updateOrderToDelivered(input: OrderDeliveredInput): Promise<Order> {
+    const order = await OrderModel.findOne({ _id: input._id })
+
+    if (!order) {
+      throw new NotFoundError('Order not found')
+    }
```
Comment (nathan-torquato) [inline backend/graphql/service/order.service.ts]:
```suggestion
  const order = await this.findSingleOrder({ _id: input._id })
```

File: backend/graphql/service/order.service.ts [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+    return await order.save()
+  }
+
+  async updateOrderToCanceled(input: OrderCanceledInput): Promise<Order> {
+    const order = await OrderModel.findOne({ _id: input._id })
+
+    if (!order) {
+      throw new NotFoundError('Order not found')
+    }
+
```
Comment (nathan-torquato) [inline backend/graphql/service/order.service.ts:45]: Same suggestion as above

File: backend/graphql/service/order.service.ts [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+    return await order.save()
+  }
+
+  async getMyOrders(context: Context): Promise<Order[]> {
+    const user = context.user
+
+    if (!user) {
+      throw new NotFoundError('User not found')
+    }
```
Comment (nathan-torquato) [inline backend/graphql/service/order.service.ts]: doesn't the `@IsLoggedIn` auth guard solve this issue?

File: backend/graphql/types/item.type.ts [hunk 7]
```diff
@@ -0,0 +1,13 @@
+import { Field, ObjectType } from 'type-graphql'
+
+@ObjectType()
+export class Item {
```
Comment (nathan-torquato) [inline backend/graphql/types/item.type.ts]: Let's name this `OrderItem`

File: backend/graphql/usecases/create-order.use-case.ts [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+      )
+
+      return {
+        productId: product!._id,
+        price: product!.price,
+        quantity: item.quantity
+      }
+    })
+
+    const totalPrice = orderItems
```
Comment (nathan-torquato) [inline backend/graphql/usecases/create-order.use-case.ts]: this returns a string, shouldn't we store it as a number?

File: backend/graphql/usecases/create-order.use-case.ts [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+import { NotFoundError } from '../errors/notFoundError'
+import { type CreateOrderInput } from '../inputs/order.input'
+import { type Context } from '../types/context.type'
+import { type Order } from '../types/order.type'
+
+export class CreateOrderUseCase {
+  async execute(input: CreateOrderInput, context: Context): Promise<Order> {
+    const user = context.user
+
+    if (!user) {
```
Comment (nathan-torquato) [inline backend/graphql/usecases/create-order.use-case.ts]: same question as above

File: backend/graphql/usecases/create-order.use-case.ts [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+      throw new NotFoundError('User not found')
+    }
+
```
Comment (nathan-torquato) [inline backend/graphql/usecases/create-order.use-case.ts]: - you don't need to replicate Mongo's pattern of `_id`
- this is actually a list of ids so it should be named as `productsIds` or somthing like that

File: backend/graphql/usecases/create-order.use-case.ts [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+    const _id = input.items.map((item) => item.productId)
+
+    const products = await ProductModel.find({
+      _id
+    })
+
+    if (products.length !== input.items.length) {
+      throw new NotFoundError('Some products not found')
```
Comment (nathan-torquato) [inline backend/graphql/usecases/create-order.use-case.ts]: let's be more specific and throw an error with the `id` that has no `Product`

**Review by nathan-torquato — CHANGES_REQUESTED** [review]
see comments

File: backend/graphql/service/order.service.ts (hunk 3 above)
Comment (Mauricio-Natacci) [reply backend/graphql/service/order.service.ts:9]: Eslint rule : 
All imports in the declaration are only used as types. Use `import type`.

File: backend/graphql/resolvers/order.resolver.ts (hunk 2 above)
Comment (Mauricio-Natacci) [reply backend/graphql/resolvers/order.resolver.ts:14]: 🚀

File: backend/database/models/order.model.ts (hunk 1 above)
Comment (Mauricio-Natacci) [reply backend/database/models/order.model.ts:4]: set to solve the "setting the mixed property" problem

File: backend/graphql/service/order.service.ts (hunk 4 above)
Comment (Mauricio-Natacci) [reply backend/graphql/service/order.service.ts]: Property 'save' does not exist on type 'Order'.

File: backend/graphql/service/order.service.ts (hunk 6 above)
Comment (Mauricio-Natacci) [reply backend/graphql/service/order.service.ts]: changed it to " const user = context.user! " and deleted if check.
in the context type user can be null. ( user: User | null )

File: backend/graphql/types/item.type.ts (hunk 7 above)
Comment (Mauricio-Natacci) [reply backend/graphql/types/item.type.ts]: done

File: backend/graphql/usecases/create-order.use-case.ts (hunk 8 above)
Comment (Mauricio-Natacci) [reply backend/graphql/usecases/create-order.use-case.ts]: you're right! but IDK why the result was right on the database. but I added Number() around it

File: backend/graphql/service/order.service.ts (hunk 4 above)
Comment (nathan-torquato) [reply backend/graphql/service/order.service.ts]: What? that makes no sense.. it's literally the same code...

**Review by nathan-torquato — APPROVED** [review]

**Pull request merged**